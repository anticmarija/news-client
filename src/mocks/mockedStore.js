import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
export const initialState = {
  topNews: {},
  countries: {
    supportedCountries: [{ countryId: "gb", countryName: "Gb" }],
    activeCountry: "gb",
    isChangingCountriesEnabled: true,
  },
  categories: {
    allCategoryArticles: [
      { title: "article1", description: "some desc" },
      { title: "article2", description: "some desc" },
    ],
    categoryArticles: {
      business: [
        { title: "article1", description: "some desc" },
        { title: "article2", description: "some desc" },
      ],
      sports: [
        { title: "article3", description: "some desc" },
        { title: "article4", description: "some desc" },
      ],
    },
  },
  article: {
    activeArticle: {
      title: "active",
      urlToImage:
        "https://i.dailymail.co.uk/1s/2020/06/17/16/29719992-0-image-a-23_1592408042536.jpg",
      description: "desc",
    },
  },
};

export const mockedStore = (state = initialState) => {
  const mockStore = configureStore(middlewares);
  return mockStore(state);
};
