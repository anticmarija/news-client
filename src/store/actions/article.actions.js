export const SET_ACTIVE_ARTICLE = "SET_ACTIVE_ARTICLE";

/**
 * Action creators
 */
export const setActiveArticle = (article) => {
  return { type: SET_ACTIVE_ARTICLE, payload: article };
};
