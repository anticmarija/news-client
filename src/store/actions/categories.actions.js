export const SET_CATEGORIES = "SET_CATEGORIES";
/**
 * Action creators
 */
export const setCategories = (categories) => {
  return { type: SET_CATEGORIES, payload: categories };
};
