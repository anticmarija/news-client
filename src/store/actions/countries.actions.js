export const SET_COUNTRIES = "SET_COUNTRIES";
export const SET_ACTIVE_COUNTRY = "SET_ACTIVE_COUNTRY";
/**
 * Action creators
 */
export const setCountries = (countries) => {
  return { type: SET_COUNTRIES, payload: countries };
};

export const setActiveCountry = (country) => {
  return { type: SET_ACTIVE_COUNTRY, payload: country };
};
