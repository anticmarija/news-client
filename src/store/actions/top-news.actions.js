import config from "../../config";

export const FETCH_ARTICLES_START = "FETCH_ARTICLES_START";
export const FETCH_ARTICLES_SUCCESS = "FETCH_ARTICLES_SUCCESS";
export const FETCH_ARTICLES_FAIL = "FETCH_ARTICLES_FAIL";

/**
 * Action creators
 */
export const fetchArticlesStart = () => {
  return { type: FETCH_ARTICLES_START };
};

export const fetchArticlesSuccess = (articles) => {
  return { type: FETCH_ARTICLES_SUCCESS, payload: articles };
};

export const fetchArticlesFail = (error) => {
  return { type: FETCH_ARTICLES_FAIL, payload: error };
};

/**
 * Fetches ARTICLES and dispatches success and error actions
 */
export const fetchArticles = (countryId) => {
  return async (dispatch) => {
    try {
      dispatch(fetchArticlesStart());

      const response = await fetch(
        `${config.getApiBaseUrl()}/top-headlines?country=${countryId}&apiKey=${config.getApiKey()}`
      );

      const { articles } = await response.json();

      dispatch(fetchArticlesSuccess(articles));
    } catch (err) {
      dispatch(fetchArticlesFail(err));
    }
  };
};
