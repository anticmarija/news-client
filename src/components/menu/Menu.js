import React from "react";
import { useLocation } from "react-router-dom";
import { useSelector, useDispatch } from "react-redux";
import { findNewRoute } from "../../utils/helpers";

import {
  MenuWrapperStyle,
  MenuItemsWrapperStyle,
  MenuCountriesWrapperStyle,
  MenuItemStyle,
  LinkStyle,
} from "./Menu.style";
import { setActiveCountry } from "../../store/actions/countries.actions";

const Menu = () => {
  const {
    supportedCountries,
    activeCountry,
    isChangingCountriesEnabled,
  } = useSelector((state) => state.countries);
  const { pathname } = useLocation();

  const dispatch = useDispatch();

  const handleChangeCountry = (countryId) => {
    if (isChangingCountriesEnabled) {
      dispatch(setActiveCountry(countryId));
    }
  };

  return (
    <MenuWrapperStyle>
      <MenuItemsWrapperStyle>
        <MenuItemStyle
          enabled
          active={pathname === `/country/${activeCountry}/`}
        >
          <LinkStyle exact to={`/country/${activeCountry}/`}>
            Top news
          </LinkStyle>
        </MenuItemStyle>
        <MenuItemStyle
          enabled
          active={pathname === `/country/${activeCountry}/categories`}
        >
          <LinkStyle exact to={`/country/${activeCountry}/categories`}>
            Categories
          </LinkStyle>
        </MenuItemStyle>
        <MenuItemStyle
          enabled
          active={pathname === `/country/${activeCountry}/search`}
        >
          <LinkStyle exact to={`/country/${activeCountry}/search`}>
            Search
          </LinkStyle>
        </MenuItemStyle>
      </MenuItemsWrapperStyle>

      <MenuCountriesWrapperStyle>
        {supportedCountries.map(({ countryId }) => (
          <MenuItemStyle
            enabled={isChangingCountriesEnabled}
            active={activeCountry === countryId}
            key={countryId}
          >
            <LinkStyle
              data-testid="country-link"
              onClick={() => handleChangeCountry(countryId)}
              active={activeCountry === countryId ? "active" : ""}
              to={
                isChangingCountriesEnabled
                  ? findNewRoute(pathname, countryId)
                  : "#"
              }
            >
              {countryId.toUpperCase() + " "}
            </LinkStyle>
          </MenuItemStyle>
        ))}
      </MenuCountriesWrapperStyle>
    </MenuWrapperStyle>
  );
};

export default Menu;
