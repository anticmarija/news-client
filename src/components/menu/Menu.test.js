import React from "react";
import Menu from "./Menu";
import { mockedStore } from "../../mocks/mockedStore";
import { Provider } from "react-redux";
import { render, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import * as actions from "../../store/actions/countries.actions";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (s) => s }),
}));

const store = mockedStore();

describe("Test Menu component", () => {
  it("should render correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Menu />
        </MemoryRouter>
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should handleChangeCountry() correctly", () => {
    const { queryByTestId } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Menu />
        </MemoryRouter>
      </Provider>
    );

    const countryLink = queryByTestId("country-link");

    const actionSpy = jest.spyOn(actions, "setActiveCountry");

    fireEvent.click(countryLink);

    expect(actionSpy).toBeCalled();
  });
});
