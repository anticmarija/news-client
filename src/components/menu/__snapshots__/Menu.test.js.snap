// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Menu component should render correctly 1`] = `
<div
  class="Menustyle__MenuWrapperStyle-sc-1neq1ee-0 kTCkyP"
>
  <div
    class="Menustyle__MenuItemsWrapperStyle-sc-1neq1ee-1 eEekmm"
  >
    <div
      class="Menustyle__MenuItemStyle-sc-1neq1ee-2 iJVOUw"
    >
      <a
        class="Menustyle__LinkStyle-sc-1neq1ee-3 ccUwVd"
        href="/country/gb/"
      >
        Top news
      </a>
    </div>
    <div
      class="Menustyle__MenuItemStyle-sc-1neq1ee-2 iJVOUw"
    >
      <a
        class="Menustyle__LinkStyle-sc-1neq1ee-3 ccUwVd"
        href="/country/gb/categories"
      >
        Categories
      </a>
    </div>
    <div
      class="Menustyle__MenuItemStyle-sc-1neq1ee-2 iJVOUw"
    >
      <a
        class="Menustyle__LinkStyle-sc-1neq1ee-3 ccUwVd"
        href="/country/gb/search"
      >
        Search
      </a>
    </div>
  </div>
  <div
    class="Menustyle__MenuCountriesWrapperStyle-sc-1neq1ee-4 hvGaFl"
  >
    <div
      class="Menustyle__MenuItemStyle-sc-1neq1ee-2 cVYhuA"
    >
      <a
        active="active"
        class="Menustyle__LinkStyle-sc-1neq1ee-3 ccUwVd"
        data-testid="country-link"
        href="/gb/"
      >
        GB 
      </a>
    </div>
  </div>
</div>
`;
